src/basic/tmpfile-util.c

```diff
@@ -28,7 +28,16 @@ int fopen_temporary_at(int dir_fd, const char *path, FILE **ret_file, char **ret
         assert(dir_fd >= 0 || dir_fd == AT_FDCWD);
 
         if (path) {
-                r = tempfn_random(path, NULL, &t);
+                struct stat st;
+
+                r = RET_NERRNO(fstatat(dir_fd, path, &st, 0));
+                if (r < 0 && r != -ENOENT)
+                        return r;
+
+                if (r >= 0 && S_ISDIR(st.st_mode))
+                        r = tempfn_random_child(path, NULL, &t);
+                else
+                        r = tempfn_random(path, NULL, &t);
```

Comment:
i find this a bit too magic, it's a security issue: if an attacker knows our program wants to create file "foo" somewhere, and we normally generate an unguessable name from it, so that the attacker cannot guess it after all you now give the attacke a really powerful weapon: it can now affect where you will create the file.

and what's worse, you even follow symlinks with the fstatat(). this is hence a way to allow creation of files at arbitrary places for an attacker

Reply:
so what I am basically saying: there should be a paramater passed in (or maybe two different function prototypes) that select whether to create the tempfile under the specified name, or below it. i.e. control of this must be by the caller